File: src/leetCodeProblems/BruteForce/CanPlaceFlowers605.java
package leetCodeProblems.BruteForce;

/**
 * LeetCode - https://leetcode.com/problems/can-place-flowers/submissions/
 *
 * TimeComplexity - O(n)
 * SpaceComplexity - O(1)
 */
public class CanPlaceFlowers605 {

    public boolean canPlaceFlowers(int[] flowerbed, int n) {

        int availablePositions = 0;

        for (int i=0; i < flowerbed.length; i++) {

            if (flowerbed[i] == 0) {

                boolean emptyLeft = false;
                boolean emptyRight = false;

                if (i == 0) {
                    emptyLeft = true;
                } else if (flowerbed[i-1] != 1) {
                    emptyLeft = true;
                }

                if (i == flowerbed.length-1) {
                    emptyRight = true;
                }
                else if (flowerbed[i+1] != 1) {
                    emptyRight = true;
                }

                //System.out.println("emptyRight ->" + emptyRight);

                if (emptyLeft && emptyRight) {
                    availablePositions++;
                    i++;
                }
            }
        }

        if (availablePositions >= n) {
            return true;
        }

        return false;
    }

    public static void main(String[] args) {

        //int[] flowerbed = {1,0,0,0,1};
        //int n = 1;

        //int[] flowerbed = {1,0,0,0,0,1};
        //int n = 2;

        //int[] flowerbed = {1,0,0,0,1,0,0};
        //int n = 2;

        int[] flowerbed = {0,0,1,0,0};
        int n = 1;

        CanPlaceFlowers605 obj = new CanPlaceFlowers605();
        System.out.println(obj.canPlaceFlowers(flowerbed, n));

    }
}

File: src/leetCodeProblems/Greedy/CanPlaceFlowers605.go
package main

/*
- Leetcode - https://leetcode.com/problems/can-place-flowers/solutions/4002498/go-simple-greedy-70-faster/
- Time - O(n)
- Space - O(1)
*/

func canPlaceFlowers(flowerbed []int, n int) bool {

	i := 0

	for i < len(flowerbed) {
		if flowerbed[i] == 0 &&
			(i+1 >= len(flowerbed) || flowerbed[i+1] == 0) &&
			(i == 0 || flowerbed[i-1] == 0) {
			n--
			flowerbed[i] = 1
			i = i + 2
		} else {
			i++
		}

		if n == 0 {
			break
		}
	}
	return n == 0
}

// func main() {
// 	// log.Println(canPlaceFlowers([]int{1, 0, 0, 0, 1}, 1))
// 	// log.Println(canPlaceFlowers([]int{1, 0, 0, 0, 1}, 2))
// 	log.Println(canPlaceFlowers([]int{1, 0, 0, 0, 0, 0, 1}, 2))
// 	log.Println(canPlaceFlowers([]int{0, 0, 1, 0, 0}, 1))

// }
